internal/core/software_renderer.rs

```diff
@@ -1024,7 +1040,9 @@ impl<'a, T: ProcessScene> crate::item_rendering::ItemRenderer for SceneBuilder<'
             let border = rect.border_width();
             let radius = rect.border_radius();
             // FIXME: gradients
-            let color = rect.background().color();
+            let color = self.alpha_color(&rect.background());
+            let border_color = self.alpha_color(&rect.border_color());
```

Comment:
Note: querying properties add a dependency node in the property engine, which allocates.
The previous code only did that if there was a radius or a border width.
Arguably, it is unlikely that none of these property are set for  BorderRectangle, so it shouldn't change much, but this would still be optimized.
```suggestion
            let border_color = if border.get() > 0.01 { self.alpha_color(&rect.border_color()) } else { Color::default() };
```